content/blog/2016/2016-08-29-sauce-pipeline.adoc

```diff
@@ -0,0 +1,271 @@
+I was able to write a pipeline that ran five tests on one browser via Sauce Connect:
+
+[source,groovy]
+----
+node {
+    stage "Build"
+    checkout scm
+
+    sh 'npm install' // <1>
+
+    stage "Test"
+    sauce('f0a6b8ad-ce30-4cba-bf9a-95afbc470a8a') { // <2>
+        sauceconnect(options: '', useGeneratedTunnelIdentifier: false, verboseLogging: false) { // <3>
+            sh './node_modules/.bin/nightwatch -e chrome --test tests/guineaPig.js || true' // <4>
+            junit 'reports/**' // <5>
+            step([$class: 'SauceOnDemandTestPublisher']) // <6>
+        }
+    }
+}
+----
+<1> Install dependencies
+<2> Use my sauce credentials
+<3> Start up the SauceConnect tunnel to Sauce Labs
+<4> Run Nightwatch.js
+<5> Use JUnit to track results and show trend graph
+<6> Link result details from Sauce Labs
+
+[NOTE]
+====
+This pipeline expects to be run from a `Jenkinsfile` in SCM.
+To copy and paste it directly into a Jenkins Pipeline job, replace the `checkout scm` step with
+`git url:'https://github.com/bitwiseman/JS-Nightwatch.js', branch: 'sauce-pipeline'`.
+====
+
+I ran this job a few times to get the JUnit report to show a trend graph.
+
+image::/images/post-images/2016-08-29/pipeline-report.png[Pipeline Report for "sauce-pipeline" Branch, role="center"]
+
+This sample app generates the `SauceOnDemandSessionID` for each test, enabling the Jenkins Sauce OnDemand Plugin's result publisher to link results to details Sauce Labs captured during the run.
+
+image::/images/post-images/2016-08-29/pipeline-results.png[Sauce Test Results, role="center"]
+
+image::/images/post-images/2016-08-29/sauce-details.png[Sauce Test Result Details, role="center"]
+
+
+=== Adding Platforms
+Next I added some more platforms to my matrix.
+This is a perfect example of the power of pipeline as code.
+Adding new platforms required changing both the pipeline and the Nightwatch.js configuration file.
+Since my pipeline was now checked in, I could change both as part of one commit and be assured my pipeline and test framework wouldn't get out of sync.
+
+I added three new environments to `nightwatch.json`:
+
+[source,js]
+----
+"test_settings" : {
+  "default": { /*----8<----8<----8<----*/ },
+  "chrome": { /*----8<----8<----8<----*/ },
+
+  "firefox": {
+    "desiredCapabilities": {
+      "platform": "linux",
+      "browserName": "firefox",
+      "version": "latest"
+    }
+  },
+  "ie": {
+    "desiredCapabilities": {
+      "platform": "Windows 10",
+      "browserName": "internet explorer",
+      "version": "latest"
+    }
+  },
+  "edge": {
+    "desiredCapabilities": {
+      "platform": "Windows 10",
+      "browserName": "MicrosoftEdge",
+      "version": "latest"
+    }
+  }
+}
+----
+
+And I modified my `Jenkinsfile` to call them:
+
+[source,groovy]
+----
+//----8<----8<----8<----8<----8<----8<----
+sauceconnect(options: '', useGeneratedTunnelIdentifier: false, verboseLogging: false) {
+    def configs = [ // <1>
```

Comment:
It's not entirely clear why it was necessary to define a list of browsers in the Jenkinsfile and in the nightwatch configuration file. Could you add a sentence explaining that somewhere around here?